sycl/test/basic_tests/id.cpp

```diff
@@ -88,37 +88,155 @@ int main() {
+   * Where OP is: ==, !=.
+   * Common by-value semantics.
+   * T must be equality comparable on the host application and within SYCL
+   * kernel functions. Equality between two instances of T (i.e. a == b) must be
+   * true if the value of all members are equal and non-equality between two
+   * instances of T (i.e. a != b) must be true if the value of any members are
+   * not equal, unless either instance has become invalidated by a move
+   * operation. Where T is id<dimensions>. */
+  {
+#define firstOneValue 10
+#define secondOneValue 19
+#define firstTwoValue 15
+#define secondTwoValue 12
+#define firstThreeValue 3
+#define secondThreeValue 22
+
+    cl::sycl::id<1> one_dim_op_one(firstOneValue);
+    cl::sycl::id<1> one_dim_op_two(secondOneValue);
+    cl::sycl::id<1> one_dim_op_another_one(firstOneValue);
+
+    cl::sycl::id<2> two_dim_op_one(firstOneValue, firstTwoValue);
+    cl::sycl::id<2> two_dim_op_two(secondOneValue, secondTwoValue);
+    cl::sycl::id<2> two_dim_op_another_one(firstOneValue, firstTwoValue);
+
+    cl::sycl::id<3> three_dim_op_one(firstOneValue, firstTwoValue,
+                                     firstThreeValue);
+    cl::sycl::id<3> three_dim_op_two(secondOneValue, secondTwoValue,
+                                     secondThreeValue);
+    cl::sycl::id<3> three_dim_op_another_one(firstOneValue, firstTwoValue,
+                                             firstThreeValue);
+
+    // OP : ==
+    // id<1> == id<1>
+    assert((one_dim_op_one == one_dim_op_two) ==
+           (firstOneValue == secondOneValue));
+    assert((one_dim_op_one == one_dim_op_another_one) ==
+           (firstOneValue == firstOneValue));
+    // id<2> == id<2>
+    assert((two_dim_op_one == two_dim_op_two) ==
+           ((firstOneValue == secondOneValue) &&
+            (firstTwoValue == secondTwoValue)));
+    assert(
+        (two_dim_op_one == two_dim_op_another_one) ==
+        ((firstOneValue == firstOneValue) && (firstTwoValue == firstTwoValue)));
+    // id<3> == id<3>
+    assert((three_dim_op_one == three_dim_op_two) ==
+           ((firstOneValue == secondOneValue) &&
+            (firstTwoValue == secondTwoValue) &&
+            (firstThreeValue == secondThreeValue)));
+    assert((three_dim_op_one == three_dim_op_another_one) ==
+           ((firstOneValue == firstOneValue) &&
+            (firstTwoValue == firstTwoValue) &&
+            (firstThreeValue == firstThreeValue)));
+    // id<1> == size_t
+    assert((one_dim_op_one == secondOneValue) ==
+           (firstOneValue == secondOneValue));
+    assert((one_dim_op_one == firstOneValue) ==
+           (firstOneValue == firstOneValue));
+
+#ifndef __SYCL_DISABLE_ID_TO_INT_CONV__
+    // size_t == id<1>
+    assert((firstOneValue == one_dim_op_two) ==
+           (firstOneValue == secondOneValue));
+    assert((firstOneValue == one_dim_op_another_one) ==
+           (firstOneValue == firstOneValue));
+#endif // __SYCL_DISABLE_ID_TO_INT_CONV__
```

Comment:
This one seems to work even with `__SYCL_DISABLE_ID_TO_INT_CONV__`, am I missing something?

Reply:
No, this should not work. 

In case #define __SYCL_DISABLE_ID_TO_INT_CONV__ : 
The next case can be compilled:
id<1> == size_t
because in this case the constructor id<1>(size_t ) used and it is similar to:
id<1> == id<1> 
and then the equality operator of array<1> used.
The test for this case is id.cpp:148 and id.cpp:184

The next case is not correct:
size_t == id<1>
there is not way to cast id to integral
this test case is under #ifndef, because it is valid only when id to size_t conversion enabled